spring-cloud-gcp-pubsub-stream-binder/src/test/java/com/google/cloud/spring/stream/binder/pubsub/config/PubSubBinderConfigurationTests.java

```diff
@@ -0,0 +1,131 @@
+import com.google.cloud.spring.pubsub.core.PubSubTemplate;
+import com.google.cloud.spring.stream.binder.pubsub.provisioning.PubSubChannelProvisioner;
+import org.junit.Test;
+
+import org.springframework.boot.autoconfigure.AutoConfigurations;
+import org.springframework.boot.test.context.runner.ApplicationContextRunner;
+import org.springframework.context.annotation.Bean;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+
+/**
+ * Tests for PubSubBinderConfiguration and its interaction with other autoconfiguration classes.
+ */
+public class PubSubBinderConfigurationTests {
+
+	@Test
+	public void testEverythingEnabled_implicit() {
+		ApplicationContextRunner baseContext = new ApplicationContextRunner()
+				// no property values specified
+				.withConfiguration(AutoConfigurations.of(
+						GcpContextAutoConfiguration.class,
+						GcpPubSubAutoConfiguration.class,
+						PubSubBinderConfiguration.class))
+				.withUserConfiguration(TestConfiguration.class);
+		baseContext.run(ctx -> assertThat(ctx).hasSingleBean(PubSubChannelProvisioner.class));
+	}
+
+	@Test
+	public void testEverythingEnabled_explicit() {
+		ApplicationContextRunner baseContext = new ApplicationContextRunner()
+				.withPropertyValues(
+						// including 'core' for completeness, but it's not super relevant to the test.
+						"spring.cloud.gcp.core.enabled=true",
+						"spring.cloud.gcp.pubsub.enabled=true",
+						"spring.cloud.gcp.pubsub.binder.enabled=true")
+				.withConfiguration(AutoConfigurations.of(
+						GcpContextAutoConfiguration.class,
+						GcpPubSubAutoConfiguration.class,
+						PubSubBinderConfiguration.class))
+				.withUserConfiguration(TestConfiguration.class);
+		baseContext.run(ctx -> assertThat(ctx).hasSingleBean(PubSubChannelProvisioner.class));
+	}
+
+	@Test
+	public void testBinderDisabled() {
+		ApplicationContextRunner baseContext = new ApplicationContextRunner()
+				.withPropertyValues(
+						"spring.cloud.gcp.pubsub.enabled=true",
+						"spring.cloud.gcp.pubsub.binder.enabled=false")
+				.withConfiguration(AutoConfigurations.of(
+						GcpContextAutoConfiguration.class,
+						GcpPubSubAutoConfiguration.class,
+						PubSubBinderConfiguration.class))
+				.withUserConfiguration(TestConfiguration.class);
+		baseContext.run(ctx -> assertThat(ctx).doesNotHaveBean(PubSubChannelProvisioner.class));
+	}
+
+	@Test
+	public void testPubSubDisabled_noReplacements() {
+		ApplicationContextRunner baseContext = new ApplicationContextRunner()
+				.withPropertyValues(
+						"spring.cloud.gcp.pubsub.enabled=false",
+						"spring.cloud.gcp.pubsub.binder.enabled=true")
+				.withConfiguration(AutoConfigurations.of(
+						GcpContextAutoConfiguration.class,
+						GcpPubSubAutoConfiguration.class,
+						PubSubBinderConfiguration.class))
+				.withUserConfiguration(TestConfiguration.class);
+		baseContext.run(ctx -> assertThat(ctx).doesNotHaveBean(PubSubChannelProvisioner.class));
```

Comment:
For this test, is it possible to try get the Binder bean in the `baseContext.run(ctx -> ...)` from the context and then checking if an exception is thrown? Maybe this is what you want to check. I think if you never try retrieving/injecting a bean that misses dependencies in the app, then you will never run into the error since it won't try to create the bean.

But I think in practice if you can get the exception to throw when the bean is retrieved/injected somewhere in the app, then this is good enough; it is how it works in other dependency injection systems too like Guice I think.

Reply:
AFAIK that's not the case w Spring - beans are created even if never injected.

Reply:
Ah I see.

Reply:
Maybe I am missing something, I'm not quite following why the test doesn't throw an exception then. Doesn't `PubSubBinderConfiguration` create the `PubSubChannelProvisioner` bean which should throw an error if `pubsub.enabled=false`?